File: C++/sorting/heapsort.cpp
#include <algorithm>
#include "heapsort.h"

#include <iostream>

void Heapsort::Sort(std::vector<int>& arr) {
  BuildHeap(arr);

  for (int i = arr.size() - 1; i >= 0; --i) {
    std::swap(arr[0], arr[i]);
    Heapify(arr, 0, i);
  }
}

void Heapsort::Heapify(std::vector<int>& arr, int index, int delimiter) {
  int left = 2 * index + 1;
  int right = 2 * index + 2;
  int largest = index;

  if (left < delimiter && arr[left] > arr[largest])
    largest = left;

  if (right < delimiter && arr[right] > arr[largest])
    largest = right;

  if (largest != index) {
    std::swap(arr[index], arr[largest]);
    Heapify(arr, largest, delimiter);
  }
}

void Heapsort::BuildHeap(std::vector<int>& arr) {
  int len = arr.size();
  for (int i = len / 2 - 1; i >= 0; --i) {
    Heapify(arr, i, len);
  }
}

File: Java/sorting/Heapsort.java
package sorting;

import java.util.List;

public class Heapsort {
  public static void sort(List<Integer> arr) {
    int len = arr.size();

    for (int i = len / 2 - 1; i >= 0; i--) {
      heapify(arr, i, len);
    }

    for (int i = len - 1; i >= 0; i--) {
      swap(arr, 0, i);
      heapify(arr, 0, i);
    }
  }

  private static void heapify(List<Integer> arr, int index, int delimiter) {
    int left = 2 * index + 1;
    int right = 2 * index + 2;
    int largest = index;

    if (left < delimiter && arr.get(left) > arr.get(largest)) largest = left;

    if (right < delimiter && arr.get(right) > arr.get(largest)) largest = right;

    if (index != largest) {
      swap(arr, index, largest);
      heapify(arr, largest, delimiter);
    }
  }

  private static void swap(List<Integer> arr, int left, int right) {
    int temp = arr.get(left);
    arr.set(left, arr.get(right));
    arr.set(right, temp);
  };
}
